Re-enable ecc-vscc and ercc-vscc

mbrandenburger opened the pull request:
This PR fixes #250 and re-enables ecc and ercc custom validation. 



<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our code of conduct and contributor guidelines: 
     https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/CONTRIBUTING.md
     https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/CODE_OF_CONDUCT.md
   In particular pay attention to the git workflows
      https://docs.google.com/document/d/1sR7YV3pSYN3NEFiW-2fUqtpsJeJrpC0EWUVtEm0Blcg/edit#heading=h.kwcug3pkefak
2. Fill out below sections.
3. Label the PR with the label of any component this PR touches.
4. ALso don't forget to sign your comments before submitting. 
   Github will complain if there is no DCO but it's easier if we don't have to hunt you down to fix that :-)

-->

**What this PR does / why we need it**:

This PR fixes #250 and re-enables ecc and ercc custom validation. 

- Change core.yaml to point to plugins/*-vscc.so
- Set ercc-vscc in ercc chaincode definition in our peer wrapper
- Set ecc-vscc for fpc chaincodes in integration tests
- add workaround in ercc-vscc to fix missing "ecc_" namespace prefix
  for fpc chaincodes; and fix mrenclave retrieval
- remove utils depency from ecc-vscc as it caused errors when peer
  loads the plugin
- improved debug logging in ecc-vscc
- add workaround in ecc-vscc for fix missing "ecc_" namespace prefix

**Which issue(s) this PR fixes**:
<!--
  list existing bug, feature and/or work-item which this PR addresses.
  You might also consider creating an issue first for the PR.
-->
Fixes #250

**Special notes for your reviewer**:

**Does this PR introduce a user-facing changes and/or breaks backward compatability?**:
<!--
  If no, you can delete this section
  If yes, describe what changes and/or what breaks ..
-->
```

g2flyer commented on plugins/ecc-vscc/ecc_validation_logic.go:
```diff
@@ -233,12 +246,25 @@ func (vscc *VSCCECC) checkEnclaveEndorsement(cis *peer.ChaincodeInvocationSpec,
+		return false
+	}
+
+	// TODO do some real check if namespace is a FPC namespce.
```
Can't we solve it right now?  Couldn't we use presence of the "ecc-vscc" validation plugins?  Otherwise, let's revert to the ecc_ (or better, fpc_) prefixing of chaincode name we did earlier? (it's really ugly, but short of changing chaincode definition which we are unlikely going to do for MVP, i guess there are only more or less ugly solutions?  (Personally, the validation-plugin route seems nicer than the prefix one ...)

BTW: this issue exists also in TLCC. So if we do not address it here then we definitely need an issue and work-plan update for it!

g2flyer replied:
From discussion on today's call i guess conclusion is that (a) presence of escc-validation is proper way of identification of FPC chaincode, (b) until ERCC is refactored, we need though a workaround for the current erccc cc2cc (presumably based on the "ercc" namespace?) and (c) we need a new (MVP) issue to implement that strategy in TLCC and remove the workaround as soon as ercc is refactored?

mbrandenburger replied:
let me look into that if we can solve it right now or not. Since we are not supporting cc2cc calls with FPC there is a chance that non-fpc chaincodes wont show up in the read/writeset. Only _lifecycle and maybe ercc? but we can exclude them. Will look into that.

g2flyer replied:
You mean "non-fpc chaincodes" _except_ ercc?!  In any case, from your latest version i think you consider it resolved for the ecc (and ercc) validation plugins?    I would tend to agree (although see some comments in my review). However, what about correspondingly also fix tlcc: do we have a corresponding issue for that?

g2flyer commented on integration/deployment_test.sh:
```diff
@@ -26,9 +26,9 @@ run_test() {
-    try ${PEER_CMD} lifecycle chaincode approveformyorg -C ${CHAN_ID} --package-id ${PKG_ID} --name auction_test --version ${CC_VERS}
-    try ${PEER_CMD} lifecycle chaincode checkcommitreadiness -C ${CHAN_ID} --name auction_test --version ${CC_VERS}
-    try ${PEER_CMD} lifecycle chaincode commit -C ${CHAN_ID} --name auction_test --version ${CC_VERS}
+    try ${PEER_CMD} lifecycle chaincode approveformyorg -C ${CHAN_ID} --package-id ${PKG_ID} --name auction_test --version ${CC_VERS} -V ecc-vscc
```
Wouldn't it be better to add the validation plugin inside the peer cli?  Mid/long-term some of these plugins will change for FPC, so it would be cleaner/easier from a functional interface point of view as the interface is simpler and would be the same even if we would move to MSP/X509 style validation.  As we anyway do not allow any user-provided plugins, the peer cli of course should also barf if any plugins provided (which it currently doesn't, e.g., if you provide an endorsement plugin, which by itself is also a bit of a problem?)

mbrandenburger replied:
> Wouldn't it be better to add the validation plugin inside the peer cli? Mid/long-term some of these plugins will change for FPC, so it would be cleaner/easier from a functional interface point of view as the interface is simpler and would be the same even if we would move to MSP/X509 style validation. As we anyway do not allow any user-provided plugins, the peer cli of course should also barf if any plugins provided (which it currently doesn't, e.g., if you provide an endorsement plugin, which by itself is also a bit of a problem?)

Created a new issue for that #372

mbrandenburger commented on plugins/ecc-vscc/ecc_validation_logic.go:
```diff
@@ -27,6 +26,19 @@ import (
+
+const SEP = "."
+
+func TransformToSGX(comp, sep string) string {
```
naaaah not really ... at least we are always using it with this const SEP.

I just copied this function from [utils/utils.go](https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/utils/utils.go#L49) to break the dependency issue.

g2flyer replied:
Argh, we really have to duplicate the function and constants?! This is really ugly.  Is this really necessary? At the very least we should add comments at both places cross-referencing each other as presumably they should be kept consistent.

g2flyer replied:
btw: i agree with morgan. Given the name of the function, the SEP as argument is rather confusing and would be better handled inside the function ....

mbrandenburger replied:
ok, I refactored utils/utils.go and removed the go-shim dependency, so we can use utils.go again within ecc-vscc and ecc_enclave. This solves code duplication concerns here. 

Also removed `sep` input from util functions, even though this should be a seperate work item and does not help to solve the actual task of this PR. But if we are already touching it here - it's OK!

mbrandenburger commented on plugins/ercc-vscc/ercc_validation_logic.go:
```diff
@@ -232,13 +236,15 @@ func policyErr(err error) *commonerrors.VSCCEndorsementPolicyError {
 	for _, ns := range txRWSet.NsRwSets {
 		for _, read := range ns.KvRwSet.Reads {
-			if read.Key == MrEnclaveStateKey && strings.HasPrefix(ns.NameSpace, "ecc_") {
+			// Note that we used to use ecc_ as prefix for every fpc chaincode; however, this has been removed
```
Ok good idea! Let's remove comment here and add to commit msg.

mbrandenburger commented:
> Looks nice to me. Does CI Run this check?

Yes! Our custom vsccs are now used in integration tests.

g2flyer commented on plugins/ercc-vscc/ercc_validation_logic.go:
```diff
@@ -187,10 +186,15 @@ func (t *VSCCERCC) checkAttestation(respPayload *peer.ChaincodeAction) error {
+		// if mrenclave not already on the ledger
 		if mrenclave == nil {
-			return errors.New("mrenclave is empty")
+			// it must be in the writeset
+			mrenclave = getMrEnclaveWriteSet(eccNamespace, txRWSet)
+			if mrenclave == nil {
+				return fmt.Errorf("no MRENCLAVE on the ledger nor in write set for %s", eccNamespace)
+			}
 		}
```
Why is this actually not the only/main case?  Do we currently allow updates? Otherwise, there shouldn't be anything in the state anyway? If so, how are we doing updates securely?  I realize this plugin will go away but the validation logic should transfered to some extent to ercc itself, so this code fragment is still somewhat relevant? maybe some comments could be helpful for the future?

mbrandenburger replied:
Well, the code also considers the case where setup is already called for another enclave , in that case mrenclave may already exists on the ledger. However, I guess this peace of logic needs bigger rework as we will consume MRENCLAVE from chaincode definition and not from ledger state. Thus, I don't think we need to touch it here or add any comment.

mbrandenburger replied:
I added a pointer to this comment in #134

g2flyer commented on plugins/ecc-vscc/ecc_validation_logic.go:
```diff
@@ -121,70 +120,70 @@ func (vscc *VSCCECC) checkEnclaveEndorsement(cis *peer.ChaincodeInvocationSpec,
-			continue
-		}
+	if string(function) == "__setup" {
+		// do nothing for setup transaction as we handle this with ercc-vscc
```
Just for clarifications -- overall, a few more comments for the whole plugin validations couldn't have hurt ... -- fabric calls validation plugins always on the (complete) transactions as as soon as there is a top-level or sub (cc2cc) transacation for a chaincode associated with a custom validation plugin?  I.e., for setup transaction both validation plugins are called -- and hence above works ... -- , any other (top-level) ercc transaction involves only ercc-vscc and any other (top-level) ecc tx involves only ecc-vscc? (regarding any other top-level ercc tx, i guess all would be rejected by the plugin with non-intuitve errors but would be safe, in any case, this plugin would be obsolete soon ...)

mbrandenburger replied:
I believe the `__setup` transaction will go away with ercc refactoring and the corresponding comments on how fabric deals with cc2cc invocations are then obsolete here.

I updated the comment but I refrain from adding details on how cc2cc are handled in fabric.

g2flyer commented on plugins/ecc-vscc/ecc_validation_logic.go:
```diff
@@ -121,70 +120,70 @@ func (vscc *VSCCECC) checkEnclaveEndorsement(cis *peer.ChaincodeInvocationSpec,
 		return err
 	}
 
-	for _, ns := range txRWSet.NsRwSets {
-		logger.Debugf("Namespace %s", ns.NameSpace)
-
-		// TODO make this more flexible
-		if !strings.HasPrefix(ns.NameSpace, "ecc") {
-			continue
-		}
```
I guess this we dropped as, as discussed, by definition the ecc plugin should be configured exactly (and only) for ecc transactions?

mbrandenburger replied:
yes

g2flyer commented on plugins/ecc-vscc/ecc_validation_logic.go:
```diff
@@ -121,70 +120,70 @@ func (vscc *VSCCECC) checkEnclaveEndorsement(cis *peer.ChaincodeInvocationSpec,
+	var readset, writeset [][]byte
+	// go through the read/writeset as produced within the enclave
+	for _, ns := range txRWSet.NsRwSets {
+		if ns.NameSpace != cis.ChaincodeSpec.ChaincodeId.Name {
+			// note that the read/write set may contain other namespaces than our fpc chaincode such as lifecycle
+			// we only care about FPC chaincode namespace and therefore filter
+			continue
+		}
```
I would have expected that ercc would be the only other namespace?  Is a custom validation plugin also called on lifecycle events?  That would be kindof surprising (e.g., for chaincode spec updates updating plugins, which plugins are called?!) but more importantly, how would we make sure the corresponding transactions are properly verified!  By doing continue what would validate the corresponding path?   In fact, i doubt that lifecycle tx would come here as they usually would not have ecc attestations and hence below verification fails?

So ideally we would test that it is ercc or fail otherwise?  In any case, i think the current comment is wrong and should be updated

mbrandenburger replied:
here we iterate over the read/writeset of a transaction associated with ecc-vscc. As an ecc-enclave endorses only read/writes that are part of the fpc chaincode namespace, any other entry showing up here are not signed by the enclave, thus, must be filtered out in order to verify the enclave signature.

What are those other entries? If you look at the logs we can see that every chaincode transaction contains a readset entry to `_lifecycle`.`$chaincode_name$`. In particular, TLCC logs show it pretty well and I assume that explorer would show it as well.
It seems that this is a check that all endorsements use the same "chaincode version". Anyway, as this appears in the readset, the entry is covered by the peer endorsement which we have on top of ecc-vscc right now. So no need for additional checks here.

g2flyer replied:
oh, i see, never saw these __lifecyce values. So definitely was wrong on that.

g2flyer commented on plugins/ercc-vscc/ercc_validation_logic.go:
```diff
@@ -232,13 +236,14 @@ func policyErr(err error) *commonerrors.VSCCEndorsementPolicyError {
-	// Readsets must contain only one ecc_*.MRENCLAVE
+	// Readsets must contain only one *some_fpc_chaincode_name*.MRENCLAVE
 	eccNamespace := ""
 	for _, ns := range txRWSet.NsRwSets {
 		for _, read := range ns.KvRwSet.Reads {
-			if read.Key == MrEnclaveStateKey && strings.HasPrefix(ns.NameSpace, "ecc_") {
+			if read.Key == MrEnclaveStateKey {
+				logger.Debugf("found MRENCLAVE within namespace: %s", ns.NameSpace)
 				if eccNamespace != "" {
-					return "", fmt.Errorf("mutiple namespaces for ecc found")
+					return "", fmt.Errorf("mutiple namespaces with MRENCLAVE key found")
 				}
 				eccNamespace = ns.NameSpace
```
I guess the current code would allow wrap an ecc _setup-tx by (via cc2cc) by another tx and "hijack" the attestation and register it with a different ecc_namespace?  Per-se i don't see immediately how this leads to bad stuff but it's a bit ugly.  Ideally, i think the attestation should explicitly bind the the ecc namespace?  (Not necessarily here, but might be good to do so in the revised version? Actually, iff we implement the new registration protocol, we should do so ... :-)

mbrandenburger replied:
I linked this conversation to #134

mbrandenburger commented:
> Current version works fine in HW mode and i think it resolved the validation questions raised in the discussion. However, i think at least one comment is misleading/wrong and we should make sure the corresponding changes are also in TLCC (probably via a separate issue?). In general, for such security-sensitive code i think a bit more comments which connects the dots could be useful (although i agree that much (but not all) of the code will change with ercc refactoring ...)

I created #373 to track corresponding todos on TLCC and also addressed your concerns regarding the comments.
